keras/engine/topology.py

```diff
@@ -1317,13 +1335,50 @@ def get_output_shape_for(self, input_shape):
         elif self.mode == 'cos':
             return (input_shapes[0][0], 1)
 
-    def compute_mask(self, input, mask=None):
-        '''TODO: add mask merging support
+    def _normalize_mask_dims(self, mask, input):
+        '''Some inconsistency in how masks are passed. This method
+        normalizes all masks to be the same size as what they are masking.
         '''
-        if mask is not None and any(mask):
-            raise Exception('Merge does not support masking, ' +
-                            'but was passed an input mask: ' + str(mask))
-        return None
+        if mask is None:
+            return K.ones_like(input)
+
+        if K.ndim(mask) == K.ndim(input):
+            return K.cast(mask, K.floatx())
+
+        assert K.ndim(mask) == K.ndim(input) - 1
+        mask = K.expand_dims(mask)
+        return K.ones_like(input) * K.cast(mask, K.floatx())
+
+    def compute_mask(self, inputs, mask=None):
+        if mask is None or not any([m is not None for m in mask]):
+            return None
+
+        assert hasattr(mask, '__len__')
+        assert len(mask) == len(inputs)
+
+        if self.mode in ['sum', 'mul', 'ave']:
+            normalized = [self._normalize_mask_dims(m, i) for m, i in zip(mask, inputs)]
+            eall = K.eall(normalized)
+            if K.ndim(eall) > 2: # AND mask along feature dimension
+                return K.all(eall, axis=2, keepdims=False)
+            return eall
+        elif self.mode in ['concat']:
+            normalized = [self._normalize_mask_dims(m, i) for m, i in zip(mask, inputs)]
+            concatenated = K.concatenate(normalized, axis=self.concat_axis)
```

Comment:
I vote for option number two. That seems the most consistent with other
parts of the Keras API.